janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java

```diff
@@ -334,6 +334,7 @@ private static boolean indexAppliesTo(IndexType index, JanusGraphElement element
                     update = new IndexUpdate<>(iIndex, updateType, getIndexKey(iIndex, record), getIndexEntry(iIndex, record, relation), relation);
                 } else {
                     assert relation.valueOrNull(type)!=null;
+                    index.resetCache();
```

Comment:
I noticed that this place is called every time we added new properties which are covered by some index.
So I was thinking, maybe we should reset cache every time we add a new key to the index (i.e. update index itself) instead of every time we add a new vertex property which is covered by some mixed index. After checking it, I see that we already do it here:
https://github.com/JanusGraph/janusgraph/blob/c1b11d4b848303ae0a2df102123fe018abad18d2/janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementSystem.java#L707
Thus, it confuses me. After checking it more, I noticed that we often create a new instance of `MixedIndexTypeWrapper`. Even so, it uses the same `base` the actual instance changes. 
Checking it more, I see that each call to this method produces a new instance even if we don't need a new instance:
https://github.com/JanusGraph/janusgraph/blob/c1b11d4b848303ae0a2df102123fe018abad18d2/janusgraph-core/src/main/java/org/janusgraph/graphdb/types/vertices/JanusGraphSchemaVertex.java#L177
I tried to change the logic to actually cache the result instead of producing a new instance and all tests passed: https://github.com/porunov/janusgraph/commit/ce9191ef302093c9ae9b13ca5d13dab1cbe8e4ad (see changes in `JanusGraphSchemaVertex.java`). I can't confirm that caching is OK here because it actually changes the logic a little bit (a new instance don't have cached `fieldMap`. Thus, it will repopulate that map on the first interaction. The only question is: do we really need to repopulate that map each time or is it OK to cache it in some / all situations?).
What I see is that `asIndexType()` is called many times in `RelationTypeVertex.getKeyIndexes()` here:
https://github.com/JanusGraph/janusgraph/blob/c1b11d4b848303ae0a2df102123fe018abad18d2/janusgraph-core/src/main/java/org/janusgraph/graphdb/types/vertices/RelationTypeVertex.java#L113

The above is just unfinished research and it could potentially be somehow connected to this PR. I can't yet verify if this fix is correct or not. I mean, it definitely fixes the problem which is great but I don't know if there are any bad consequences which we get if we apply this fix (i.e. is there any performance degradation or GC pressure increase due to this fix?). This fix could probably be right I just isn't too familiar with this part of the code. Thus, couldn't verify properly. Hope someone could help verifying it.

cc @li-boxuan @rngcntr @mad

Reply:
Are you saying https://github.com/porunov/janusgraph/commit/ce9191ef302093c9ae9b13ca5d13dab1cbe8e4ad fix #1876? I don't see a test in that commit thus want to confirm.

Reply:
No, it doesn't fix the problem. That's just a different thing which I noticed

Reply:
Got it, thanks for the clarification!